### PR TITLE
Custom enumerables

### PR DESCRIPTION
**Pull request by layzrd**
ruby Custom Iterators

- add #my_each
- add #my_each_with_index 
- add #my_select ]
- add #my_all?
- add #my_any?
- add #my_none?
- add #my_count
- add #my_map
- add #my_inject

### CODE REVIEW

**Review by serdg0 — CHANGES_REQUESTED** [review]
Hi @abredi !

Congratulations, your project it's heading in the right direction!

I must say, I applaud your `my_select`, it's **identical** to the original one! Congratulations for that! :clap: :clap: 

This project's goal is to create you own enumerables as equal as the original ones, and in some cases the methods are not returning the same. I'm going to list that below:

### my_each & my_each_with_index
- [ ] When a block (or proc) is passed, it's returning an array of booleans, instead of returning the array itself after calling the given block.
- [ ] When passing a block to a range instead of an array, it should return the range itself with the block called for each element.

### my_all? & my_any & my_none?
In this methods, everytime I try to call them the following error is appearing:
`undefined local variable or method 'array' for #<Array:0x000055aa72639760>`
Inside the method you call to your `my_each` an undefined variable `array`, but you never actually declare it.
For this methods, consider the following cases: When a REGEX is passed, when a Range is passed, when a block is passed and when nothing is passed.

### my_map
- [ ] When no block is given it should return an enumerator, but I got a `LocalJumpError`.

### my_inject
- [ ] The original `inject` method has a parameter for an initial value. When I test `my_inject`, it's returning an `ArgumentError`
- [ ] The original `inject` doesn't mutate the original array. This is, it return only the intended output, and the original array is kept with no changes.
- [ ] Consider the following cases: When a block is passed, when a range is passed.

### multiply_els
- [ ] This method is working good. There is just one small thing to implement, and is, that it accepts an array as a parameter. So, instead of being called like: `array.multiply_els`, fix in a way that can be called like `multiply_els(array)`

You are encouraged to check if your methods returns the same as the original ones. This can be done in another file, or in your terminal irb. Also, don't forget to checkout the docs https://ruby-doc.org/core-2.6.5/Enumerable.html

**Happy Coding!** :computer: :nerd_face:
**_Sergio Diaz_**

**Comment by sumancrest0001** [comment]
Status: changes required
Hello Team, you have done a good job so far and made a lot of improved as comapared to the last review. But we still need to work on a couple of methods to get approval. Please consider the following cases.

**#my_each**
when a block is given when self is a range returns the range itself after calling the given block once for each element in self

**#my_all?**
Please compare your custom my_all? methods with all? method in the following cases

- when no block or argument is given returns true when none of the collection members are false or nil

- when a class is passed as an argument returns true if all of the collection is a member of such class::Integer

- when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric

- when a Regex is passed as an argument returns false if any of the collection does not match the Regex

- when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

**my_any?**
when a Regex is passed as an argument returns true if any of the collection matches the Regex

** my_none?**
when a Regex is passed as an argument returns false if any of the collection matches the Regex

**#my_inject**

- does not mutate the original array

- when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block

- when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method

**Review by sumancrest0001 — CHANGES_REQUESTED** [review]
**Kindly make requested changes and submit the project for another review.
If you have any query, message me on slack Suman Shrestha
happy coding !!!**

File: lib/main.rb [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    flag = true
+    return check_condition(flag, array, cond) if cond
+
+    return check_block(flag, array) unless block_given?
+
+    array.my_each do |item|
+      return false unless yield(item)
+    end
+    flag
+  end
```
Comment (maelfosso) [inline lib/main.rb]:
```rb
2.6.5 :012 > false_array = [1, false, 'hi', []]
2.6.5 :013 > false_array.my_all?
 => false 
2.6.5 :014 > false_array.all?
 => false 
2.6.5 :015 > true_array = [1, true, 'hi', []]
2.6.5 :016 > true_array.my_all?
 => false 
2.6.5 :017 > true_array.all?
 => true 
2.6.5 :018 >
```

Comment (maelfosso) [reply lib/main.rb]:
```rb
2.6.5 :019 > array
 => [6, 4, 0, 4, 0, 1, 3, 7, 5, 7, 1, 6, 2, 6, 2, 0, 8, 4, 4, 1, 7, 5, 0, 8, 2, 0, 3, 6, 7, 6, 3, 8, 4, 7, 7, 0, 4, 3, 6, 2, 5, 7, 8, 7, 5, 2, 5, 3, 0, 4, 3, 5, 7, 5, 0, 8, 5, 0, 4, 6, 5, 1, 1, 1, 2, 7, 4, 1, 2, 8, 6, 2, 0, 4, 7, 3, 3, 7, 1, 1, 2, 6, 2, 5, 7, 1, 0, 7, 3, 7, 4, 7, 6, 4, 0, 7, 6, 3, 0, 0] 
2.6.5 :020 > array.my_all?(Integer)
 => false 
2.6.5 :021 > array.all?(Integer)
 => true 
2.6.5 :022 >
```

Comment (maelfosso) [reply lib/main.rb]:
```rb
2.6.5 :024 > words
 => ["dog", "door", "rod", "blade"] 
2.6.5 :025 > words.my_all?(/d/)
 => true 
2.6.5 :026 > words.all?(/d/)
 => true 
2.6.5 :027 > words.all?(/o/)
 => false 
2.6.5 :028 > words.my_all?(/o/)
 => true 
2.6.5 :029 > 
```

Comment (maelfosso) [reply lib/main.rb]:
```rb
2.6.5 :031 > array
 => [6, 4, 0, 4, 0, 1, 3, 7, 5, 7, 1, 6, 2, 6, 2, 0, 8, 4, 4, 1, 7, 5, 0, 8, 2, 0, 3, 6, 7, 6, 3, 8, 4, 7, 7, 0, 4, 3, 6, 2, 5, 7, 8, 7, 5, 2, 5, 3, 0, 4, 3, 5, 7, 5, 0, 8, 5, 0, 4, 6, 5, 1, 1, 1, 2, 7, 4, 1, 2, 8, 6, 2, 0, 4, 7, 3, 3, 7, 1, 1, 2, 6, 2, 5, 7, 1, 0, 7, 3, 7, 4, 7, 6, 4, 0, 7, 6, 3, 0, 0] 
2.6.5 :032 > array.my_all?(3)
 => false 
2.6.5 :033 > array.all?(3)
 => false 
2.6.5 :034 > arr = []
2.6.5 :035 > 5.times { arr << 3 }
 => 5 
2.6.5 :036 > array.my_all?(3)
 => false 
2.6.5 :037 > array.all?(3)
 => false 
2.6.5 :038 > 
2.6.5 :039 > arr.my_all?(3)
 => false 
2.6.5 :040 > arr.all?(3)
 => true 
2.6.5 :041 > 
```

File: lib/main.rb [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    return check_condition(flag, array, cond) if cond
+
+    return check_block(flag, array) unless block_given?
+
+    array.my_each do |item|
+      return true if yield(item)
+    end
+    flag
+  end
```
Comment (maelfosso) [inline lib/main.rb:118]:
```rb
2.6.5 :043 > words
 => ["dog", "door", "rod", "blade"] 
2.6.5 :044 > words.any?(/z/)
 => false 
2.6.5 :045 > words.my_any?(/z/)
 => false 
2.6.5 :046 > words.any?(/d/)
 => true 
2.6.5 :047 > words.my_any?(/d/)
 => false 
2.6.5 :048 >
```

File: lib/main.rb [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    return check_condition(flag, array, cond) if cond
+
+    return check_block(flag, array) unless block_given?
+
+    array.my_each do |item|
+      return false if yield(item)
+    end
+    flag
+  end
```
Comment (maelfosso) [inline lib/main.rb:132]:
```rb
2.6.5 :050 > words
 => ["dog", "door", "rod", "blade"] 
2.6.5 :051 > words.none?(/z/)
 => true 
2.6.5 :052 > words.my_none?(/z/)
 => true 
2.6.5 :053 > words.none?(/d/)
 => false 
2.6.5 :054 > words.my_none?(/d/)
 => true 
2.6.5 :055 >
```

File: lib/main.rb [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+      array.shift
+    end
+
+    return check_symbol(acc, init, symbol, array) if init.is_a?(Symbol) || symbol.is_a?(Symbol)
+
+    array.my_each do |item|
+      acc = yield(acc, item)
+    end
+    acc
+  end
```
Comment (maelfosso) [inline lib/main.rb]:
```rb
2.6.5 :058 > operation = proc { |sum, n| sum + n }
2.6.5 :059 > array.my_inject(&operation)
 => 391 
2.6.5 :060 > array.inject(&operation)
 => 385 
2.6.5 :061 > 
2.6.5 :062 > array
 => [4, 0, 4, 0, 1, 3, 7, 5, 7, 1, 6, 2, 6, 2, 0, 8, 4, 4, 1, 7, 5, 0, 8, 2, 0, 3, 6, 7, 6, 3, 8, 4, 7, 7, 0, 4, 3, 6, 2, 5, 7, 8, 7, 5, 2, 5, 3, 0, 4, 3, 5, 7, 5, 0, 8, 5, 0, 4, 6, 5, 1, 1, 1, 2, 7, 4, 1, 2, 8, 6, 2, 0, 4, 7, 3, 3, 7, 1, 1, 2, 6, 2, 5, 7, 1, 0, 7, 3, 7, 4, 7, 6, 4, 0, 7, 6, 3, 0, 0] 
2.6.5 :063 > actual = range.my_inject(4) { |prod, n| prod * n }
2.6.5 :064 > expected = range.inject(4) { |prod, n| prod * n }
2.6.5 :065 > actual == expected
 => true 
2.6.5 :066 > 
2.6.5 :067 > range
 => 5..50 
2.6.5 :068 >
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thank you for applying all the suggested changes.
Please, a bit of effort to close it.

Kindly take a look on the comments left below and then make the suggested changes. When done, submit a new code review request.

Thanks
Mael FOSSO

File: lib/main.rb (hunk 1 above)
Comment (layzrd) [reply lib/main.rb]: too long for the main point. anyhow fixed!

**Review by christianotieno — APPROVED** [review]
# PROJECT APPROVED :champagne:
Excellent job on this project. :fireworks: Keep it up.
##
***All the required tasks have been fulfilled in this project. I will go ahead and mark this as complete.***
- Congratulations on the splendid job @abredi :+1:
- You can now merge these changes.
-  I wish you all the very best in your next tasks 🥇 
##
Happy coding :smile:
**[Christian Otieno](https://github.com/ChristianOtieno)**
[Twitter](https://twitter.com/iamchrisotieno) [linkedin](https://www.linkedin.com/in/christianotieno/) [GitLab](https://gitlab.com/christianotieno)

**Pull request merged**